File: app/src/main/java/fi/kaupunkifillarit/KaupunkifillaritApplication.java
package fi.kaupunkifillarit;

import android.content.Context;
import android.support.multidex.MultiDexApplication;

import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.answers.Answers;

import fi.kaupunkifillarit.inject.KaupunkifillaritComponent;
import io.fabric.sdk.android.Fabric;

public class KaupunkifillaritApplication extends MultiDexApplication {
    protected KaupunkifillaritComponent component;

    @Override
    public void onCreate() {
        super.onCreate();

        if (!BuildConfig.DEBUG) {
            Fabric.with(this, new Crashlytics(), new Answers());
            // TODO Timber.plant(new CrashlyticsTree());
        }

        buildComponentAndInject();
    }

    protected void buildComponentAndInject() {
        component = KaupunkifillaritComponent.Initializer.init(this);
        component.inject(this);
    }

    public KaupunkifillaritComponent component() {
        return component;
    }

    public static KaupunkifillaritApplication get(Context context) {
        return (KaupunkifillaritApplication) context.getApplicationContext();
    }
}

File: app/src/main/kotlin/fi/kaupunkifillarit/KaupunkifillaritApplication.kt
package fi.kaupunkifillarit

import android.app.Application
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.answers.Answers
import com.google.android.gms.common.GoogleApiAvailability
import com.patloew.rxlocation.RxLocation
import io.fabric.sdk.android.Fabric

open class KaupunkifillaritApplication : Application() {
    val sharedPreferences: SharedPreferences by lazy {
        PreferenceManager.getDefaultSharedPreferences(this)
    }

    val googleApiAvailability: GoogleApiAvailability by lazy {
        GoogleApiAvailability.getInstance()
    }

    val rxLocation: RxLocation by lazy {
        RxLocation(this)
    }

    override fun onCreate() {
        super.onCreate()

        if (!BuildConfig.DEBUG) {
            Fabric.with(this, Crashlytics(), Answers())
        }
    }
}
